---
title: 回调地狱及其解决方法
categories: Ajax
tags: [Ajax,前端]
cover: /picture/ajax.jpg
date: 2024-03-07 17:43
---

## 同步与异步的概念

1.  [同步代码](https://developer.mozilla.org/zh-CN/docs/Learn/JavaScript/Asynchronous/Introducing "同步代码")：逐行执行，需原地等待结果后，才继续向下执行
2.  [异步代码](https://developer.mozilla.org/zh-CN/docs/Learn/JavaScript/Asynchronous/Introducing "异步代码")：调用后耗时，不阻塞代码继续执行（不必原地等待），在将来完成后触发回调函数传递结果
3.  回答代码打印顺序：发现异步代码接收结果，使用的都是回调函数

```javascript
const result = 0 + 1
console.log(result)
setTimeout(() => {
  console.log(2)
}, 2000)
document.querySelector('.btn').addEventListener('click', () => {
  console.log(3)
})
document.body.style.backgroundColor = 'pink'
console.log(4)
```

> 结果：1， 4， 2

> 按钮点击一次打印一次 3

## 回调地狱

### 概念

1.  需求：展示默认第一个省，第一个城市，第一个地区在下拉菜单中

![](image_OVqQWgqW9J.png)

1.  概念：在回调函数中嵌套回调函数，一直嵌套下去就形成了回调函数地狱
2.  缺点：可读性差，异常无法捕获，耦合性严重，牵一发动全身

```javascript
axios({ url: 'http://hmajax.itheima.net/api/province' }).then(result => {
  const pname = result.data.list[0]
  document.querySelector('.province').innerHTML = pname
  // 获取第一个省份默认下属的第一个城市名字
  axios({ url: 'http://hmajax.itheima.net/api/city', params: { pname } }).then(result => {
    const cname = result.data.list[0]
    document.querySelector('.city').innerHTML = cname
    // 获取第一个城市默认下属第一个地区名字
    axios({ url: 'http://hmajax.itheima.net/api/area', params: { pname, cname } }).then(result => {
      document.querySelector('.area').innerHTML = result.data.list[0]
    })
  })
})
```

## Promise链式调用

1.  概念：依靠 then() 方法会返回一个新生成的 Promise 对象特性，继续串联下一环任务，直到结束
2.  细节：then() 回调函数中的返回值，会影响新生成的 Promise 对象最终状态和结果
3.  好处：通过链式调用，解决回调函数嵌套问题

![](image_4Ctns6OsAO.png)

```javascript
/**
 * 需求：把省市的嵌套结构，改成链式调用的线性结构
*/
// 1. 创建Promise对象-模拟请求省份名字
const p = new Promise((resolve, reject) => {
  setTimeout(() => {
    resolve('北京市')
  }, 2000)
})

// 2. 获取省份名字
const p2 = p.then(result => {
  console.log(result)
  // 3. 创建Promise对象-模拟请求城市名字
  // return Promise对象最终状态和结果，影响到新的Promise对象
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve(result + '--- 北京')
    }, 2000)
  })
})

// 4. 获取城市名字
p2.then(result => {
  console.log(result)
})

// then()原地的结果是一个新的Promise对象
console.log(p2 === p)//false
```

## 使用promise解决回调地狱

1.  目标：使用 Promise 链式调用，解决回调函数地狱问题
2.  做法：每个 Promise 对象中管理一个异步任务，用 then 返回 Promise 对象，串联起来

![](image_zNsMQ5t9hh.png)

```javascript
/**
 * 目标：把回调函数嵌套代码，改成Promise链式调用结构
 * 需求：获取默认第一个省，第一个市，第一个地区并展示在下拉菜单中
*/
let pname = ''
// 1. 得到-获取省份Promise对象
axios({url: 'http://hmajax.itheima.net/api/province'}).then(result => {
  pname = result.data.list[0]
  document.querySelector('.province').innerHTML = pname
  // 2. 得到-获取城市Promise对象
  return axios({url: 'http://hmajax.itheima.net/api/city', params: { pname }})
}).then(result => {
  const cname = result.data.list[0]
  document.querySelector('.city').innerHTML = cname
  // 3. 得到-获取地区Promise对象
  return axios({url: 'http://hmajax.itheima.net/api/area', params: { pname, cname }})
}).then(result => {
  console.log(result)
  const areaName = result.data.list[0]
  document.querySelector('.area').innerHTML = areaName
})
```

事实上，promise是callback风格的一个语法糖，它通过实现链式调用的方式来将回调函数的嵌套扁平化来达到解决回调地狱的目的。

往回调函数里按照成功和失败的类别分别传入两个回调函数，就可以分别处理成功和失败的结果。

```javascript
function foo(handleSucceed, handleFailed) {
  if(isMistake) {
    return handleFailed(new Error('It`s a mistake'));
  } else {
    return handleSucceed('You got it!');
  }
}

foo(function (result) {
  console.log(result);
}, function (error) {
  console.error(error);
});
```

与promise实现的版本进行对比：

```javascript
const pms = new Promise((resolve, reject) => {
  // do sth.
  if(isMistake) {
    return reject(new Error('It`s a mistake'));
  } else {
    return resolve('You got it!');
  }
});

pms.then(result => {
  console.log(result);
}).catch(error => {
  console.error(error);
});
```

会发现实质上，这里的`resolve`和`reject`正是两个回调函数，就如同前面一个例子里面的`handleSucceed`和`handleFailed`一样。而这两个回调函数的传入方式，从上一个例子的直接两个参数传入，变成了通过`then`方法和`catch`方法来进行传入。

相比而言，`Promise`的方式更加语义化，更容易理解——给主流程留下一个承诺，在之后可以通过该承诺获得子流程的执行结果。

## async和await函数解决回调地狱

1.  概念：在 async 函数内，使用 await 关键字取代 then 函数，等待获取 Promise 对象成功状态的结果值。
2.  async关键字作为一个关键字放到申明函数前面，表示该函数为要给异步任务，不会阻塞后面函数的执行。`async`函数其实是一个返回值为`Promise`对象的函数。或者更准确地说：`async`关键字声明的函数被调用时，实质上是创建了一个`Promise`对象。
3.  `await`则表示：执行后面的表达式——不论是异步还是同步的表达式——并获取到`resolve`的结果，如果执行出错，则抛出到外层`async`函数的`catch`回调中。`await`同时作为`Promise`中的`resolve`和`reject`工作了。
4.  做法：使用 async 和 await 解决回调地狱问题

```javascript
/**
 * 目标：掌握async和await语法，解决回调函数地狱
 * 概念：在async函数内，使用await关键字，获取Promise对象"成功状态"结果值
 * 注意：await必须用在async修饰的函数内（await会阻止"异步函数内"代码继续执行，原地等待结果）
*/
// 1. 定义async修饰函数
async function getData() {
  // 2. await等待Promise对象成功的结果
  const pObj = await axios({url: 'http://hmajax.itheima.net/api/province'})
  const pname = pObj.data.list[0]
  const cObj = await axios({url: 'http://hmajax.itheima.net/api/city', params: { pname }})
  const cname = cObj.data.list[0]
  const aObj = await axios({url: 'http://hmajax.itheima.net/api/area', params: { pname, cname }})
  const areaName = aObj.data.list[0]


  document.querySelector('.province').innerHTML = pname
  document.querySelector('.city').innerHTML = cname
  document.querySelector('.area').innerHTML = areaName
}

getData()
```

## async 函数和 await 捕获错误

try 和 catch 的作用：语句标记要尝试的语句块，并指定一个出现异常时抛出的响应

```javascript
try {
  // 要执行的代码
} catch (error) {
  // error 接收的是，错误消息
  // try 里代码，如果有错误，直接进入这里执行
}

```

尝试把代码中 url 地址写错，观察 try catch 的捕获错误信息能力

```javascript
/**
 * 目标：async和await_错误捕获
*/
async function getData() {
  // 1. try包裹可能产生错误的代码
  try {
    const pObj = await axios({ url: 'http://hmajax.itheima.net/api/province' })
    const pname = pObj.data.list[0]
    const cObj = await axios({ url: 'http://hmajax.itheima.net/api/city', params: { pname } })
    const cname = cObj.data.list[0]
    const aObj = await axios({ url: 'http://hmajax.itheima.net/api/area', params: { pname, cname } })
    const areaName = aObj.data.list[0]

    document.querySelector('.province').innerHTML = pname
    document.querySelector('.city').innerHTML = cname
    document.querySelector('.area').innerHTML = areaName
  } catch (error) {
    // 2. 接着调用catch块，接收错误信息
    // 如果try里某行代码报错后，try中剩余的代码不会执行了
    console.dir(error)
  }
}

getData()
```

`async`函数其实是`Promise`的语法糖。并且它通过形似同步编程的形式，达成了前述`Promise`的链式调用，并以此解决回调地狱的问题。
